### PR TITLE
Resource partitioner

### PR DESCRIPTION
**Pull request by hkaiser**

### CODE REVIEW

**Comment by biddisco** [comment]
This cannot be merged until I (or you @hkaiser) eliminates and fixes the commit marked "WIP: FIX THIS PROPERLY"

**Comment by biddisco** [comment]
@shoshijak - Look! your work has not been wasted and might be merged soon. 
@aserio - please send a ste||ar t-shirt to shoshana when this branch goes into master.
@hkaiser - I am fine with all of this but have not tested the last couple of commits that address oversubscription and dynamic thread pools. I will look at them ASAP

**Comment by biddisco** [comment]
Could I please ask @KhalidHasanov to have a look through this branch and check if the work he contributed to the throttling policy has been accidentally deleted. If this is the case, then I will either edit the commit that deleted it and force push, or if @KhalidHasanov can provide a patch that reinstates anything missing, we can merge it back in. Thanks

**Comment by hkaiser** [comment]
@KhalidHasanov, @sithhell, @biddisco I think this branch allows to remove the throttling policy altogether as it now can be held outside of the core HPX libraries. This scheduler is very project specific, so let's leave it in project space.

**Comment by biddisco** [comment]
Very good point. @KhalidHasanov you should have a look at the file
```
examples/resource_partitioner/simple_resource_partitioner.cpp
```
on this branch (around line 277) to see how to supply a custom scheduler to the default pool. 
This allows you to package your scheduler in user code, rather than making it part of the main hpx distribution. At some point, we will probably go through the schedulers that are currently available and remove or clean up some of them.

**Comment by KhalidHasanov** [comment]
@biddisco Thanks for pointing out the example file. I will have a look once the branch is merged and see how it goes.

**Comment by sithhell** [comment]
Only a few minor comments here. This PR is rather big and we should get it in ASAP, and keep improving it as we gain experience with the new feature.

File: examples/resource_partitioner/shared_priority_scheduler.hpp [hunk 1]
```diff
@@ -0,0 +1,1308 @@
...
+              , numa_sensitive_(numa_sensitive)
+              , numa_pinned_(numa_pinned)
+              , initialized_(false)
+            {
+                LOG_CUSTOM_MSG("Constructing scheduler with num queues "
+                    << decnumber(num_worker_queues));
+                victim_queues_.clear();
+                victim_queues_.resize(num_worker_queues);
+                //
+                BOOST_ASSERT(num_worker_queues != 0);
```
Comment (hkaiser) [inline examples/resource_partitioner/shared_priority_scheduler.hpp]: Good catch, thanks. We should add that to the obsolete macro tests in inspect.

**Comment by hkaiser** [comment]
> This PR is rather big and we should get it in ASAP, and keep improving it as we gain experience with the new feature.

I couldn't agree more.

File: tests/unit/parallel/segmented_algorithms/partitioned_vector_target.cpp [hunk 2]
```diff
@@ -12,6 +12,18 @@
...
+             && HPX_HAVE_MAX_CPU_COUNT > 64)
+ #  if defined(HPX_HAVE_MAX_CPU_COUNT)
+ #    include <bitset>
+ #    include <hpx/runtime/serialization/bitset.hpp>
+ #  else
+ #    include <boost/dynamic_bitset.hpp>
+ #    include <hpx/runtime/serialization/dynamic_bitset.hpp>
+ #  endif
+ #endif
+
```
Comment (hkaiser) [inline tests/unit/parallel/segmented_algorithms/partitioned_vector_target.cpp]: Yes, this should go away. It's a leftover from Shoshana's experiments...

**Comment by hkaiser** [comment]
@shoshijak Thanks again for all your work on this! @aserio will get in contact with you to send a STE||AR-Group t-shirt as a small token of appreciation of your work.

**Pull request merged**